Slurm scripts patches from condor (SOFTWARE-3587)

edquist opened the pull request:
There were some changes to the `slurm_*` scripts in the htcondor repo that didn't make their way back to BLAH: cherry-pick them in here.

Some of these upstream commits added patches for source files not maintained in upstream htcondor; these are applied directly here to the BLAH sources.

JaimeFrey committed:
Remove debugging statements.

JaimeFrey committed:
Let the user set the slurm partition to use. #5780

Map the blahp's batch queue name option to slurm's partition name
option.

JaimeFrey committed:
Add blahp options for project name and runtime limit. #6451

These are two common job attributes in the batch systems supported by
the blahp. These attributes are specific to to the batch type under the
grid universe. BatchProject is a string that's used with the project or
allocation option of the batch system's submission interface.
BatchRuntime is an integer that limits the job's wallclock execution
time, expressed in seconds.

---

This is a cherry-pick of cbab688f1e from htcondor; everything other
than the slurm_* script changes are from the runtime-project.patch
introduced in that commit.

JaimeFrey committed:
Don't assume default shell is bash in blahp scripts. #6560

In a few places, the blahp scripts use the 'source' command with the
default shell. On Debian and Ubuntu, the default shell is dash, which
doesn't have that command. Change these places to invoke bash
specifically.

---

This is bash-source.patch cherry-picked from upstream condor 639c65cbc7

timtheisen committed:
(#6841) Update comparisions to None

Comparisons to singletons like None should always be done with
is or is not, never the equality operators.

JaimeFrey committed:
Support Slurm multi-cluster configurations in the grid universe. #7167

For Slurm, the batch Queue job parameter to the blahp can now include
an optional '@clustername' suffix to specify an alternate cluster that
the job should use. The job id generated for the blahp will include the
cluster name to allow subsequent actions to target the correct cluster.

---

cherry-picked from 58c06d2b2f in upstream htcondor

edquist commented:
@djw8605 and @brianhlin - this pulls over various changes that were missed from the `slurm_*` scripts that were separately maintained in the upstream htcondor repo.  I believe we've got everything now.  (The files do not match though since we've also made changes in the osg version.)

Anyway, feel free to look this over before I merge this.

(And remember don't hit the merge button, since for now we still want to fetch and merge to the UW AFS repo instead, which mirrors back to github.)

djw8605 commented:
Also looks like it edits the other schedulers as well.  All seems good to me.  We need to remember the `BatchQueue` command for hosted-ce's and CE routes.   Probably a very important change.

edquist commented:
>Also looks like it edits the other schedulers as well.

Correct; yeah for instance in [this commit](https://github.com/htcondor/htcondor/commit/cbab688f1eddd8a95ace1f341b8c9163378c4446) (from the condor repo) they made changes to both the slurm scripts maintained there, and they added [this patch](https://github.com/htcondor/htcondor/blob/cbab688f1eddd8a95ace1f341b8c9163378c4446/externals/bundles/blahp/1.16.5.1/runtime-project.patch) which made changes for other blahp stuff not maintained directly within the condor repo.  They seemed to go together so I applied those patches directly to the blahp sources as part of the cherry-picking.

brianhlin reviewed (dismissed):
Seems like there's some duplicate logic for the Slurm partition

https://github.com/htcondor/BLAH/blob/c45c8d3837abdd63f24d2e82f0b55928b8b581f6/src/scripts/slurm_submit.sh#L65-L66

edquist commented:
What's the duplicate part?

brianhlin commented:
That line and this line:

https://github.com/htcondor/BLAH/blob/c45c8d3837abdd63f24d2e82f0b55928b8b581f6/src/scripts/slurm_submit.sh#L81

`-p` is the same as `--partition`

edquist commented:
Ah, good catch; yeah looks like Jaime and Derek each added a partition option independently.

djw8605 commented:
Great minds... and all that.

edquist committed:
drop duplicate slurm partition option

note we drop this first one as the second occurrence since the second
one potentially gets an updated $bls_opt_queue

edquist commented:
I removed one of duplicate partition options, keeping the one added by Jaime, since that one potentially gets an updated `$bls_opt_queue`

brianhlin commented:
I do really like these changes where we actually make walltime requests first class. One thing that we should think about, is that some of the local submit attributes scripts may have references to walltime. What happens for those batch systems if the walltime request is set twice? Do we need to remove anything from the default submit attrs scripts that we ship? Are we going to have to let users know that they should remove their own custom walltime requests in their submit attrs scripts?

I'm not entirely sure I understand the `BatchProject` stuff, it seems to be for accounting for allocations?

edquist commented:
Yeah the commit message for 73fe1af9ca5 talks a bit about `BatchProject`, but I don't understand it real well myself.

edquist commented:
Ooh.  I guess @timtheisen gets mentioned for "pushing" / "merging" the commit, since the automated pushes are done with his github account.

brianhlin commented:
> Do we need to remove anything from the default submit attrs scripts that we ship?

@edquist could you check the local submit attrs files to see if any of them are setting walltime requests?

edquist commented:
@brianhlin, I think it is, [here](https://github.com/htcondor/BLAH/blob/devel/src/scripts/slurm_local_submit_attributes.sh#L23-L28).  Is this going to cause a problem?

brianhlin commented:
@djw8605 do you know what happens if there are multiple `#SBATCH --time` directives set? I think the right thing to do here is just remove those contents in the local submit attrs script

djw8605 commented:
I just ran a test, the last `--time` command is used.  The other is ignored.

brianhlin commented:
Thanks @djw8605! @edquist I think the appropriate thing to do would be to keep the directives in the local submit attributes for now since they're only set if `Walltime` is set in `remote_cequirements`. Then @timtheisen can note in the 8.9.x release notes that the method is deprecated in favor of using `BatchRuntime`
